airflow/models/dag.py

```diff
@@ -1093,6 +1093,21 @@ def add_task(self, task):
         :param task: the task you want to add
         :type task: task
         """
+        if task.task_id in self.task_dict:
+            # TODO: raise an error in Airflow 2.0
+            warnings.warn(
+                'The requested task could not be added to the DAG because a '
+                'task with task_id {} is already in the DAG. Starting in '
+                'Airflow 2.0, trying to overwrite a task will raise an '
+                'exception.'.format(task.task_id),
+                category=PendingDeprecationWarning)
+        else:
+            self.task_dict[task.task_id] = task
+            task.dag = self
+
+            if hasattr(task.__init__, "reapply_defaults"):
+                task.__init__.reapply_defaults(self)
```

Comment:
Interesting. Didn't know that the wrapper can be accessed in this way.

Reply:
Functions can have properties just like normal objects in python.

```python

def x():
    pass

x.a = 1
```

That's all we're doing here (we put the `reapply_defaults` attribute inside the decorator which closes over the original `kwargs`, so we can get this behaviour we want.